File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/intonation/view/IntonationView.kt
package com.musicmuni.voxatrace.demo.sections.intonation.view

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.musicmuni.voxatrace.accura.model.IntonationAnalysisResult
import com.musicmuni.voxatrace.accura.model.PitchingScore
import com.musicmuni.voxatrace.demo.sections.intonation.viewmodel.IntonationViewModel

/**
 * Intonation Analysis View - pitch accuracy against Equal Temperament and Just Intonation.
 */
@Composable
fun IntonationView(viewModel: IntonationViewModel = viewModel()) {
    val context = LocalContext.current

    val eqResult by viewModel.eqResult.collectAsStateWithLifecycle()
    val jiResult by viewModel.jiResult.collectAsStateWithLifecycle()
    val eqScore by viewModel.eqScore.collectAsStateWithLifecycle()
    val jiScore by viewModel.jiScore.collectAsStateWithLifecycle()
    val isAnalyzing by viewModel.isAnalyzing.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "Intonation Analysis",
            style = MaterialTheme.typography.titleMedium
        )

        Text(
            text = "Analyze pitch accuracy against Equal Temperament and Just Intonation",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        OutlinedButton(
            onClick = { viewModel.analyzeOffline(context) },
            enabled = !isAnalyzing
        ) {
            Text("Analyze Alankaar Voice")
        }

        if (isAnalyzing) {
            Box(
                modifier = Modifier.fillMaxWidth(),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator()
            }
        }

        // Equal Temperament results
        if (eqResult != null || eqScore != null) {
            IntonationResultSection(
                title = "Equal Temperament",
                result = eqResult,
                score = eqScore
            )
        }

        // Just Intonation results
        if (jiResult != null || jiScore != null) {
            IntonationResultSection(
                title = "Just Intonation",
                result = jiResult,
                score = jiScore
            )
        }

        ApiInfoCard()
    }
}

@Composable
private fun IntonationResultSection(
    title: String,
    result: IntonationAnalysisResult?,
    score: PitchingScore?
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant
        )
    ) {
        Column(
            modifier = Modifier.padding(12.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.SemiBold
            )

            if (score != null) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Column {
                        Text(
                            text = "Score",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Text(
                            text = "%.0f".format(score.score),
                            style = MaterialTheme.typography.titleLarge,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }

                    Column(horizontalAlignment = Alignment.End) {
                        Text(
                            text = "Swaras Detected",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Text(
                            text = "${score.swaraCount}",
                            style = MaterialTheme.typography.titleLarge,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }

            // Per-swara table
            val swaras = result?.swaras
            if (!swaras.isNullOrEmpty()) {
                HorizontalDivider()

                // Header row
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text(
                        text = "Swara",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier.weight(1f)
                    )
                    Text(
                        text = "Deviation",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier.weight(1f)
                    )
                    Text(
                        text = "Remark",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier.weight(1f)
                    )
                }

                swaras.forEach { swara ->
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween
                    ) {
                        Text(
                            text = swara.label,
                            style = MaterialTheme.typography.bodySmall,
                            modifier = Modifier.weight(1f)
                        )
                        Text(
                            text = "%+.1f cents".format(swara.deviationCents),
                            style = MaterialTheme.typography.bodySmall,
                            modifier = Modifier.weight(1f)
                        )
                        Text(
                            text = swara.deviationRemark,
                            style = MaterialTheme.typography.bodySmall,
                            modifier = Modifier.weight(1f)
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun ApiInfoCard() {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.tertiaryContainer.copy(alpha = 0.3f)
        )
    ) {
        Column(
            modifier = Modifier.padding(8.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Text(
                text = "APIs Demonstrated:",
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.Medium
            )
            Text(
                text = "Accura.analyzePitching() - intonation analysis (EQ / JI)",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = "Accura.calculateScore() - overall pitching score",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = "PitchDetection.createContourExtractor() - pitch extraction",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = "SonixDecoder.decode() - audio file loading",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Intonation/IntonationView.swift
import SwiftUI
import VoxaTrace

/// Intonation analysis view demonstrating Accura for pitch accuracy measurement.
struct IntonationView: View {
    @StateObject private var viewModel = IntonationViewModel()

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Intonation Analysis")
                .font(.headline)

            Text("Analyze pitch accuracy against Equal Temperament and Just Intonation")
                .font(.subheadline)
                .foregroundColor(.secondary)

            Button("Analyze Alankaar Voice") {
                viewModel.analyzeOffline()
            }
            .buttonStyle(.borderedProminent)
            .disabled(viewModel.isAnalyzing)

            if viewModel.isAnalyzing {
                ProgressView()
                    .frame(maxWidth: .infinity)
            }

            if let eqResult = viewModel.eqResult {
                resultSection(
                    title: "Equal Temperament",
                    result: eqResult,
                    score: viewModel.eqScore
                )
            }

            if let jiResult = viewModel.jiResult {
                resultSection(
                    title: "Just Intonation",
                    result: jiResult,
                    score: viewModel.jiScore
                )
            }

            apiInfoCard
        }
    }

    // MARK: - Result Section

    private func resultSection(
        title: String,
        result: IntonationAnalysisResult,
        score: PitchingScore?
    ) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.subheadline)
                .fontWeight(.semibold)

            HStack {
                VStack(alignment: .leading) {
                    Text("Score")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    if let score = score {
                        Text(String(format: "%.0f", score.score))
                            .font(.title2)
                            .fontWeight(.bold)
                            .foregroundColor(scoreColor(score.score))
                    } else {
                        Text("--")
                            .font(.title2)
                    }
                }

                Spacer()

                VStack(alignment: .trailing) {
                    Text("Notes")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    Text("\(result.notes.count)")
                        .font(.title2)
                        .fontWeight(.bold)
                }
            }

            // Per-note rows
            ForEach(Array(result.notes.enumerated()), id: \.offset) { _, note in
                noteRow(note)
            }
        }
        .padding(12)
        .background(Color(.secondarySystemBackground))
        .cornerRadius(8)
    }

    private func noteRow(_ note: NoteAnalysis) -> some View {
        HStack {
            Text(note.label)
                .font(.subheadline)
                .fontWeight(.medium)
                .frame(width: 40, alignment: .leading)

            Text(String(format: "%+.1f cents", note.deviationCents))
                .font(.caption)
                .foregroundColor(.secondary)

            Spacer()

            Text(String(format: "%.0f", note.score))
                .font(.caption)
                .foregroundColor(scoreColor(note.score))
        }
    }

    // MARK: - API Info Card

    private var apiInfoCard: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("APIs Demonstrated:")
                .font(.caption)
                .fontWeight(.medium)

            Text("Accura.analyzePitching() - intonation analysis (EQ/JI)")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("Accura.calculateScore() - overall pitch accuracy score")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("PitchDetection.createContourExtractor() - pitch extraction")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("SonixDecoder.decode() - audio file loading")
                .font(.caption2)
                .foregroundColor(.secondary)
        }
        .padding(8)
        .background(Color(.tertiarySystemBackground))
        .cornerRadius(6)
    }

    // MARK: - Helpers

    private func scoreColor(_ score: Float) -> Color {
        if score >= 70 { return .green }
        if score >= 40 { return .orange }
        return .red
    }
}

/// Backward compatibility alias.
typealias IntonationSection = IntonationView

#Preview {
    IntonationView()
        .padding()
}
